dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/citation/AbstractCitationFormatsConverter.java

```diff
@@ -33,54 +31,75 @@ public CitationBuilder(boolean escapeHtml) {
 
         // -------------------- LOGIC --------------------
 
-        public CitationBuilder add(String... values) {
-            elements.addAll(Arrays.asList(values));
+        public CitationBuilder add(String value) {
+            partElements.add(new Token(value, Token.Type.STATIC));
             return this;
         }
 
-        public CitationBuilder addFormatted(String value) {
-            return addFormatted(value, "");
+        public CitationBuilder value(String value) {
+            if (isNotBlank(value)) {
+                partElements.add(new Token(escapeHtml ? StringEscapeUtils.escapeHtml4(value) : value, Token.Type.DATA));
+            } else {
+                partElements.add(new Token(StringUtils.EMPTY, Token.Type.DATA));
+            }
+            return this;
         }
 
-        public CitationBuilder addFormatted(String value, String delimiter) {
-            return addFormatted(value, delimiter, delimiter);
+        public CitationBuilder rawValue(String value) {
+            partElements.add(new Token(value, Token.Type.DATA));
+            return this;
         }
 
-        public CitationBuilder addFormatted(String value, String startDelimiter, String endDelimiter) {
-            if (isNotEmpty(value)) {
-                elements.add(startDelimiter
-                        + (escapeHtml ? StringEscapeUtils.escapeHtml4(value) : value)
-                        + endDelimiter);
+        public CitationBuilder urlValue(String text, String url) {
+            if (text == null) {
+                partElements.add(new Token(StringUtils.EMPTY, Token.Type.DATA));
+            } else {
+                String value = escapeHtml && url != null
+                        ? "<a href=\"" + url + "\" target=\"_blank\">" + StringEscapeUtils.escapeHtml4(text) + "</a>"
+                        : text;
+                partElements.add(new Token(value, Token.Type.DATA));
             }
             return this;
         }
 
-        public String escapeHtmlIfNeeded(String value) {
-            if (isNotEmpty(value)) {
-                return escapeHtml ? StringEscapeUtils.escapeHtml4(value) : value;
-            }
-            return null;
+        public CitationBuilder endPart() {
+            return endPart(", ");
         }
 
-        public String formatURL(String text, String url) {
-            if (text == null) {
-                return null;
+        public CitationBuilder endPart(String delimiter) {
+            if (partElements.stream()
+                    .filter(t -> Token.Type.DATA == t.type)
+                    .map(t -> t.value)
+                    .allMatch(StringUtils::isNotBlank)) {
```

Comment:
Is there any difference between `partElements.stream().filter(t -> Token.Type.DATA == t.type).count() > 0` and in `value`, `rawValue` and `urlValue` just not adding blank values to the list? Just wondering

Reply:
There is one subtle difference: the current code allows to create line consisting of only static strings (it's even used in Rds, although only once if I remember correctly), and with the code from comment this would be not possible without checking for additional condition.

Reply:
Ok. Thanks for clarification.